moVirt/src/main/java/org/ovirt/mobile/movirt/rest/OVirtClient.java

```diff
@@ -682,6 +670,65 @@ public void setupVersionHeader() {
         restClient.setHeader("Version", authenticator.getApiMajorVersion());
     }
 
+    private void setPersistentV3AuthHeaders() {
+        restClient.setHeader(SESSION_TTL, "120"); // 2h
+        restClient.setHeader(PREFER, "persistent-auth, csrf-protection");
+    }
+
+    private void resetClientSettings() {
+        restClient.setHeader(SESSION_TTL, "");
+        restClient.setHeader(PREFER, "");
+        restClient.setCookie(JSESSIONID, "");
+        restClient.setAuthentication(new HttpAuthentication() {
+            @Override
+            public String getHeaderValue() {
+                // empty authentication - e.g. not the basic one
+                return "";
+            }
+        });
+    }
+
+    private void updateClientBeforeCall() {
+        restClient.setHeader(FILTER, Boolean.toString(!authenticator.hasAdminPermissions()));
+        requestFactory.setCertificateHandlingMode(authenticator.getCertHandlingStrategy());
+        restClient.setRootUrl(authenticator.getApiUrl());
+    }
+
+    /**
+     * @param username username
+     * @param password password
+     * @return auth token depending on API version, if enforceBasicAuth is true then returns empty string
+     */
+    public String login(String username, String password) {
+        resetClientSettings();
+        updateClientBeforeCall();
+        restClient.setHttpBasicAuth(username, password);
+
+        Api api = restClient.loginV3();
```

Comment:
I think it is better to separate these two concerns. First you should try to login, first with OAuth, and if that fails, then try with Basic authentication. Once you are logged in (no matter which authentication mechanism) you can check and decide what version of the API to use.

Any suggestion to improve this?

Note that the old authentication mechanism, Basic auth, will be removed when v3 is removed, probably with version 4.1 of the engine.